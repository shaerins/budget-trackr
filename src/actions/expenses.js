import { v4 as uuidv4 } from 'uuid';
import database from "../firebase/firebase";

// add expense 
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});

// start add expense
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "",
            note = "",
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref("expenses").push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// remove expense
export const removeExpense = ({ id } = {}) => ({
    type: "REMOVE_EXPENSE",
    id
});

// start remove expense
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        });
    };
};

// edit expense
export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});

// start edit expense
export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// set expenses
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
});

//start set expenses
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref("expenses").once("value").then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    };
};