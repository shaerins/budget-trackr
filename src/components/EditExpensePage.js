import React from "react";

const EditExpensePage = () => (
    <div>
        edit expenses page
    </div>
);

export default EditExpensePage;