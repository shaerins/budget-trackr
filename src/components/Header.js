import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active" exact={true}>Create</NavLink>
        <NavLink to="/Edit" activeClassName="is-active" exact={true}>Edit</NavLink>
        <NavLink to="/help" activeClassName="is-active" exact={true}>Help</NavLink>
    </header>
);

export default Header;