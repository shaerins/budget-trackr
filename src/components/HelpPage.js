import React from "react";

const HelpPage = () => (
    <>
        help page
    </>
);

export default HelpPage;