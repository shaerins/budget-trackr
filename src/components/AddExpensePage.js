import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";
import React from "react";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push("/");
    };
    render() {
        return (
            <>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit} />
            </>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);