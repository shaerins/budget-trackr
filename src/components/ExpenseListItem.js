import React from "react";
import { removeExpense } from "../actions/expenses";
import { connect } from "react-redux";

const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => (
    <div>
        <h3>{description}</h3>
        <p>Amount: ${amount} - Date: {createdAt}
            <button onClick={() => {
                dispatch(removeExpense({id}))
            }}
            >Remove</button>
        </p>
    </div >
);

export default connect()(ExpenseListItem);