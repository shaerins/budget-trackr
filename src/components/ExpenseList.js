import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectedExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-on-mobile">Expenses</div>
            <div className="show-on-desktop">Expense</div>
            <div className="show-on-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <span className="list-item list-item__message">You have no expenses.</span>
                ) : (
                        props.expenses.map((expense) => {
                            return <ExpenseListItem key={expense.id} {...expense} />
                        })
                    )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectedExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);