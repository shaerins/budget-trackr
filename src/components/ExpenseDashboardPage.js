import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
    <React.Fragment>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </React.Fragment>
);

export default ExpenseDashboardPage;