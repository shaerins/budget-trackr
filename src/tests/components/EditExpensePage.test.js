import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage editExpense={editExpense} removeExpense={removeExpense} history={history} />);
});

test("should render editexpensepage correctly", () => {
    expect(wrapper).toMatchSnapshot();
});