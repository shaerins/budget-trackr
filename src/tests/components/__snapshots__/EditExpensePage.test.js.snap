// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render editexpensepage correctly 1`] = `
<Fragment>
  <ExpenseForm
    expense={
      Object {
        "amount": 70000,
        "createdAt": -432000000,
        "description": "gas bill",
        "id": "3",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</Fragment>
`;
